services/management/actions.go

```diff
@@ -0,0 +1,217 @@
+	action, ok := a.actionsStorage.Load(actionID)
+	if !ok {
+		a.logger.Error("Unknown action with ID: %s.", actionID)
+		return
+	}
+
+	res := a.agentsRegistry.SendRequest(ctx, action.PmmAgentID, &agentpb.StopActionRequest{
+		ActionId: actionID,
+	})
+	a.logger.Infof("CancelAction response: %+v.", res)
+}
+
+// GetActionResult gets PMM Action with the given ID from action results storage.
+//nolint:unparam
+func (a *ActionsService) GetActionResult(ctx context.Context, actionID string) (models.ActionResult, bool) {
+	return a.actionsStorage.Load(actionID)
+}
+
+type preparedAction struct {
+	ID         string
+	Name       managementpb.ActionType
+	Params     []string
+	PmmAgentID string
+}
+
+func (a *ActionsService) prepareAction(rp RunActionParams) (preparedAction, error) {
+	action := preparedAction{
+		ID:         "/action_id/" + uuid.New().String(),
+		PmmAgentID: rp.PmmAgentID,
+		Name:       rp.ActionName,
+		Params:     rp.ActionParams,
+	}
+	var err error
+
+	switch action.Name {
+	case managementpb.ActionType_PT_SUMMARY:
+		action.PmmAgentID, err = findPmmAgentIDByNodeID(a.db.Querier, rp.PmmAgentID, rp.NodeID)
+
+	case managementpb.ActionType_PT_MYSQL_SUMMARY:
+		action.PmmAgentID, err = findPmmAgentIDByServiceID(a.db.Querier, rp.PmmAgentID, rp.ServiceID)
+
+	case managementpb.ActionType_MYSQL_EXPLAIN:
+		action.PmmAgentID, err = findPmmAgentIDByServiceID(a.db.Querier, rp.PmmAgentID, rp.ServiceID)
+
+	case managementpb.ActionType_ACTION_TYPE_INVALID:
+		return action, errUnsupportedAction
+	}
+
+	return action, err
+}
+
+func findPmmAgentIDByNodeID(q *reform.Querier, pmmAgentID, nodeID string) (string, error) {
+	agents, err := models.PMMAgentsForNode(q, nodeID)
+	if err != nil {
+		return "", err
+	}
+	return validatePmmAgentID(pmmAgentID, agents)
+}
+
+func findPmmAgentIDByServiceID(q *reform.Querier, pmmAgentID, serviceID string) (string, error) {
+	agents, err := models.PMMAgentsForService(q, serviceID)
+	if err != nil {
+		return "", err
+	}
+	return validatePmmAgentID(pmmAgentID, agents)
+}
+
+func validatePmmAgentID(pmmAgentID string, agents []*models.Agent) (string, error) {
+	// no explicit ID is given, and there is only one
+	if pmmAgentID == "" && len(agents) == 1 {
+		return agents[0].AgentID, nil
+	}
+
+	// no explicit ID is given, and there are zero or several
+	if pmmAgentID == "" {
+		return "", errPmmAgentIDNotFound
+	}
+
+	// check that explicit agent id is correct
+	for _, a := range agents {
+		if a.AgentID == pmmAgentID {
+			return a.AgentID, nil
+		}
+	}
+	return "", errPmmAgentIDNotFound
+}
+
+// InMemoryActionsStorage in memory action results storage.
+type InMemoryActionsStorage struct {
+	container sync.Map
+}
+
+// NewInMemoryActionsStorage created new InMemoryActionsStorage.
+func NewInMemoryActionsStorage() *InMemoryActionsStorage {
+	return &InMemoryActionsStorage{}
+}
+
+// Store stores an action result in action results storage.
+func (s *InMemoryActionsStorage) Store(result models.ActionResult) {
```

Comment:
Accept pointer.

Reply:
Can you explain why, please? As I understand there is no reason to use a pointer, isn't it? I heard a lot of speeches about `overusing` pointers in Go. Also, I heard that in many cases pointers can be expensive than passing by value in the context of performance.

Reply:
Value semantics is very useful in some areas (like concurrency) and not so much in others. In that case, it would be expected for storage update fields like CreatedAt/UpdatedAt. Change on value will be lost.

In general, it makes sense to use value semantics where we can commit to it (that means not using slices, maps, structs with them, etc.) and where it is useful (for example, where we want to avoid sharing – again, that's mostly about concurrency). Passing structs (not simpler types) by the pointer is the default.

Performance-wise, passing a single pointer is faster than passing a struct with 5 fields 4 of which (strings) contains pointers (that does not break value semantics, but on lower level pointers are there). Performance doesn't matter at all in that case, but I think it is important to understand how argument passing works.

Reply:
> I think it is important to understand how argument passing works.

- https://codeburst.io/pass-by-reference-in-go-demystified-81e0e8dfa2ad

As I understand, there is no `passing by reference` concept, so pointers don't work like in "C". When you pass a pointer, you copy a pointer variable. 

**I'll change this but I don't understand why** 😕 : 
> Passing structs (not simpler types) by the pointer is the default.

**I can't find this rule anywhere.**
A lot of sources talks differently. For example, there is some info about performance and when to use pointers:

- https://pocketgophers.com/why-use-pointers/
- https://github.com/golang/go/wiki/CodeReviewComments#pass-values
- http://goinbigdata.com/golang-pass-by-pointer-vs-pass-by-value/

>  Passing by value in Go may be significantly cheaper than passing by pointer. This happens because Go uses escape analysis to determine if variable can be safely allocated on function’s stack frame, which could be much cheaper then allocating variable on the heap. Passing by value simplifies escape analysis in Go and gives variable a better chance to be allocated on the stack.